app/jobs/set_catkeys_and_barcodes_job.rb

```diff
@@ -0,0 +1,129 @@
+  # @option params [String] :barcodes_option option to update the barcodes
+  # @option params [String] :csv_file CSV string
+  def perform(bulk_action_id, params)
+    super
+
+    update_pids, catkeys, barcodes = params_from(params)
+
+    with_bulk_action_log do |log|
+      log.puts("#{Time.current} Starting SetCatkeysAndBarcodesJob for BulkAction #{bulk_action_id}")
+      update_druid_count
+      update_pids.each_with_index do |current_druid, i|
+        change_set = ItemChangeSet.new do |change|
+          change.catkey = catkeys[i] if catkeys
+          change.barcode = barcodes[i] if barcodes
+        end
+        update_catkey_and_barcode(current_druid, change_set, log) if change_set.changed?
+      end
+      log.puts("#{Time.current} Finished SetCatkeysAndBarcodesJob for BulkAction #{bulk_action_id}")
+    end
+  end
+
+  private
+
+  def update_catkey_and_barcode(current_druid, change_set, log)
+    log.puts("#{Time.current} Beginning SetCatkeysAndBarcodesJob for #{current_druid}")
+    cocina = Dor::Services::Client.object(current_druid).find
+
+    unless ability.can?(:manage_item, cocina)
+      log.puts("#{Time.current} Not authorized for #{current_druid}")
+      return
+    end
+
+    log_update(change_set, log)
+
+    begin
+      object_client = Dor::Services::Client.object(current_druid)
+      cocina_item = object_client.find
+      state_service = StateService.new(current_druid, version: cocina_item.version)
+      open_new_version(cocina_item.externalIdentifier, cocina_item.version, version_message(change_set)) unless state_service.allows_modification?
+      ItemChangeSetPersister.update(cocina_item, change_set)
+
+      bulk_action.increment(:druid_count_success).save
+      log.puts("#{Time.current} Catkey/barcode added/updated/removed successfully")
+    rescue StandardError => e
+      log.puts("#{Time.current} Catkey/barcode failed #{e.class} #{e.message}")
+      bulk_action.increment(:druid_count_fail).save
+      nil
+    end
+  end
+
+  def dig_from_params_if_option_set(params, key)
+    params.dig(:set_catkeys_and_barcodes, key).split("\n").map(&:strip).map(&:presence) if params.dig(:set_catkeys_and_barcodes, "use_#{key}_option".to_sym) == '1'
+  end
+
+  def params_from(params)
+    if params[:csv_file]
+      update_pids = []
+      catkeys = nil
+      barcodes = nil
+      CSV.parse(params[:csv_file], headers: true).each do |row|
+        update_pids << row['Druid']
+        if row.header?('Catkey')
+          catkeys ||= []
+          catkeys << row['Catkey'].presence
+        end
+        if row.header?('Barcode')
+          barcodes ||= []
+          barcodes << row['Barcode'].presence
+        end
+      end
+    else
+      catkeys = dig_from_params_if_option_set(params, :catkeys)
+      barcodes = dig_from_params_if_option_set(params, :barcodes)
+      update_pids = pids
+    end
+    [update_pids, catkeys, barcodes]
```

Comment:
Take this with a grain of salt: as I'm looking at the code in this job, and looking at the screenshot, I am wondering if implementing this feature as two bulk actions (one for CSVs, one using textareas) might not afford both 1) more easily grokkable code due to reduced responsibilities, and 2) a more streamlined user experience. Tagging @andrewjbtw and @astridu on this. The downside, of course, is that it means yet more bulk actions.

Reply:
Does that mean instead of 1 option for bar codes, it's 2 options: "Manage catkeys/bar codes by entering manually", "Manage catkeys/bar codes by uploading CSV"?

Reply:
@astridu I think so, yeah. When we move to the new design, I think this could work as a single bulk action but with the current UI... there's a lot going on on the page.

Reply:
It's ok with me too. But I wonder if we should take this chance to segment the dropdown menu into "Bulk actions that need a CSV" and "Bulk actions that don't". I think this would have been the only one to combine the two.

Reply:
After some preliminary testing I'm now more in favor of separating the CSV from the form. I got tripped up a few times with a CSV in terms of figuring out when to use a form field and when not to. It was mostly inattention on my part (not remembering that I'd clicked "populate from search" before adding my CSV) but I can see it confusing others too.